### DISCUSSION TITLE
Make a URL analyser - full version

### DISCUSSION THREAD

**Discussion by tomcrane** [digirati-co-uk/iiif-manifest-editor#248] [Ideas]
This is discussed in the Wiki, in https://github.com/digirati-co-uk/iiif-manifest-editor/wiki/Import-of-IIIF-Resources and https://github.com/digirati-co-uk/iiif-manifest-editor/wiki/Import-of-Content-Resources

This is used anywhere the user drops a URL into the user interface when attempting to open or import something. 

It's not itself a UI, but it's called by several UI components. It's a code component. But UI components use it to sometimes present additional choices - e.g., if you're trying to import a canvas and you drop a Manifest URL in, and there's more than one canvas, the ME needs to know which one.

This _analyser_ returns a JSON object - a resource - which might be an image, with width and height properties, or an image service, or a IIIF Presentation resource.

## MVP

 - Is the URL an Image?
 - Is the URL an Image Service?
 - Is the URL an image but is actually a parameterisation of an image service?
 - Is the URL a Canvas?
 - Is the URL a Manifest?
 - Is the URL a link to an encoded state?
 - Is the string an encoded content state?

## Future

 - Any web page may be pasted and the code will try to find a manifest, or an image service, or the largest image, on that page, depending on what's needed by the caller. This eventually (in future) allows for a very forgiving content-assembly process. It can even have special handlers for Wikimedia pages and others. It can even resolve to an adaptor/proxy service which returns "legacy image pyramid" providers like Wikimedia and Flickr as level 0, `sizes`-only services.

## Image Examples

```
analyse("https://upload.wikimedia.org/wikipedia/commons/4/4d/NASA_Apollo_17_Lunar_Roving_Vehicle.jpg");

{
   "id": "https://upload.wikimedia.org/wikipedia/commons/4/4d/NASA_Apollo_17_Lunar_Roving_Vehicle.jpg",
   "type": "Image",
   "format": "image/jpeg",
   "width": 3000,
   "height": 3000
}
```

(whatever the best way to measure the image is... load into an invisible `img` tag and then ask for `.width` and `.height` ?)

```
analyse("https://iiif.wellcomecollection.org/image/b21151763_0005.jp2");
analyse("https://iiif.wellcomecollection.org/image/b21151763_0005.jp2/info.json");
analyse("https://iiif.wellcomecollection.org/image/b21151763_0005.jp2/full/1000,/0/default.jpg");

{
  "@context" : "http://iiif.io/api/image/2/context.json",
  "@id" : "https://iiif.wellcomecollection.org/image/b21151763_0005.jp2",
  "protocol" : "http://iiif.io/api/image",
  "width" : 1533,
  "height" : 2520,
  "tiles" : [
     { "width" : 256, "height" : 256, "scaleFactors" : [ 1,2,4,8,16 ] }
  ],
  "sizes" : [
{"width": 623, "height": 1024}, {"width": 243, "height": 400}, {"width": 122, "height": 200}, {"width": 61, "height": 100}
  ],
  "profile" : [
     "http://iiif.io/api/image/2/level1.json",
     {
       "formats" : [ "jpg" ],
       "qualities" : [ "native","color","gray" ],
       "supports" : ["regionByPct","sizeByForcedWh","sizeByWh","sizeAboveFull","rotationBy90s","mirroring","gray"]
     }
  ]
}
```

The difference between the third example here and the first Wikimedia image is that the analyser recognised that the last part of the path is an Image API parameterisation, and pulled out the image service.

## Other examples

Other IIIF resources are more straightforward.

```
analyse("https://iiif.wellcomecollection.org/presentation/b21151763");

// Sees that the resource is a Manifest. But does it load it into Vault and return a ref? Or just return a ref that the caller could decide to load into Vault?

{
   "id": "https://iiif.wellcomecollection.org/presentation/b21151763",
   "type": "Manifest"
}
```

```
analyse("JTdCJTIyaWQlMjIlM0ElMjJodHRwcyUzQSUyRiUyRnByZXZpZXcuaWlpZi5pbyUyRmNvb2tib29rJTJGMzMzMy1jaG9pY2UlMkZyZWNpcGUlMkYwMDMzLWNob2ljZSUyRmNhbnZhcyUyRnAxJTIzeHl3aCUzRDgzNyUyQzkxMyUyQzc3MSUyQzMzNCUyMiUyQyUyMnR5cGUlMjIlM0ElMjJDYW52YXMlMjIlMkMlMjJwYXJ0T2YlMjIlM0ElNUIlN0IlMjJpZCUyMiUzQSUyMmh0dHBzJTNBJTJGJTJGcHJldmlldy5paWlmLmlvJTJGY29va2Jvb2slMkYzMzMzLWNob2ljZSUyRnJlY2lwZSUyRjAwMzMtY2hvaWNlJTJGbWFuaWZlc3QuanNvbiUyMiUyQyUyMnR5cGUlMjIlM0ElMjJNYW5pZmVzdCUyMiU3RCU1RCU3RA");

{
    "id": "https://preview.iiif.io/cookbook/3333-choice/recipe/0033-choice/canvas/p1#xywh=837,913,771,334",
    "type": "Canvas",
    "partOf": [
        {
            "id": "https://preview.iiif.io/cookbook/3333-choice/recipe/0033-choice/manifest.json",
            "type": "Manifest"
        }
    ]
}
```

## Possible future behaviour (NOT FOR NOW!)

```
analyse("ImageService", "https://en.wikipedia.org/wiki/A_Burial_At_Ornans");
analyse("ImageService", "https://commons.wikimedia.org/wiki/File:Gustave_Courbet_-_A_Burial_at_Ornans_-_Google_Art_Project_2.jpg");

{
    "@context": "http://iiif.io/api/image/2/context.json",
    "@id": "https://iiifmediawiki.herokuapp.com/image/a/a0/Gustave_Courbet_-_A_Burial_at_Ornans_-_Google_Art_Project_2.jpg",
    "width": 6042,
    "height": 2777,
    "profile": [
        "http://iiif.io/api/image/2/level0.json"
    ],
    "protocol": "http://iiif.io/api/image",
    "sizes": [
        { "height": 147, "width": 320 },
        { "height": 294, "width": 640 },
        { "height": 367, "width": 800 },
        { "height": 470, "width": 1024 },
        { "height": 588, "width": 1280 },
        { "height": 1176, "width": 2560 },
        { "height": 2777, "width": 6042 }
    ]
}
```

```
analyse("ImageService", "https://en.wikipedia.org/wiki/A_Burial_At_Ornans");
analyse("ImageService", "https://commons.wikimedia.org/wiki/File:Gustave_Courbet_-_A_Burial_at_Ornans_-_Google_Art_Project_2.jpg");

{ 
   "id": "https://iiifmediawiki.herokuapp.com/presentation/File:Gustave_Courbet_-_A_Burial_at_Ornans_-_Google_Art_Project_2.jpg",
   "type": "Manifest"
}
```

```
analyse("Manifest", "https://collections.britishart.yale.edu/catalog/tms:1095");

// sniff out the link to a Manifest from this web page
{ 
   "id": "https://manifests.collections.yale.edu/ycba/obj/1095",
   "type": "Manifest"
}
```

**Comment by stephenwf** [comment]
Building blocks:
* [Image URL to definition](https://github.com/atlas-viewer/iiif-image-api/blob/893a5b8b4988d41d8aa47e3d05cc6843021a7a6c/src/utility/infer-size-from-url.ts)
* [Image request](https://github.com/digirati-co-uk/iiif-net/blob/main/src/IIIF/IIIF/ImageApi/ImageRequest.cs) (also ported to JS)

Potential strategies:
* Knowledge in the Vault - have I loaded this before?
* Synchronous regex or extension
* Patterned URLs (`{domain}/presentation/v2/{id}` mapped to manifest)
* HEAD requests to get headers/mime
* Byte-by-byte load EXIF and bail early without loading whole image (complex)
* Load the whole resource (ideal for JSON etc.)
* Stream partial JSON until we can identify (Oboe.js)

**Comment by tomcrane** [comment]
The URL analyser will be invoked from different places that are going to expect different things.

So there's an optional argument that specifies the type of thing you _expect_ it to be. URL analyser can use this to decide what order it's going to try stuff. e.g.,

```
analyse(myUrl, "Manifest");
```

If we expect it be a Manifest, maybe the first thing URL analyser does is load the URL into Vault.

Whereas:

```
analyse(myUrl, "Image", "ImageService");
```

...might start it on a different path, looking for Images first.

URL Analyser will still keep trying - so you could specify "Image" and still get a Manifest back if that's what the URL is. It's up to the caller to decide that it's an error. And there may be scenarios (like add content to Canvas) where you expect Image, or Audio, or whatever - but can still _handle_ a Manifest coming back.

**Comment by stephenwf** [comment]
Would this be useful as a vault helper?

**Comment by tomcrane** [comment]
Possibly. It feels like a standalone thing because it could be used elsewhere, although may need different logic in different contexts.

Whether it's focused enough to fit alongside the other Vault helpers, I'm not sure.

**Comment by stephenwf** [comment]
Having thought about this more, there is no one analysis that fits all. Depending on the context it will more likely be one thing vs. another. There is also a big difference between static analysis of the URL and probing.

So the analysis should work in 3 steps:
* **Static analysis** of the URL to find the best order to try for dynamic analysis.
* **Dynamic analysis** to determine the `Content-Type` 
* **Content analysis** to refine further and extract information

### Static analysis.

* Looking for file extensions
* Looking for structure matching image server request
* Matching other known resources
* Detecting known patterns (e.g. YouTube video)

### Dynamic analysis

* HEAD request to resource, returning content type
* Trying to fetch `/info.json` to find image server
* Bypassing CORS by using browser elements `<img />`, `<audio/>` and `<video />`
* Fetching full resource, checking content directly

### Content analysis

* Determining the width/height of an image or video
* Determining the duration of a video
* Discovering an `info.json` from an image served from IIIF Image API endpoint
* Determining the type + version of a resource (e.g. IIIF Presentation 2 Manifest)

With a skeleton framework of helpers for each of these steps, we can start to create issues that cover the detection of a specific resource that fits into one of these categories.
